### DISCUSSION TITLE
UI & UX Discussions

### DISCUSSION THREAD

**Discussion by raresvanca** [slowscript/warpinator-android#57] [General]
So I was watching some videos about the new Material U, so in my boredom, I came up with the idea to implement the design of Material and Material U respectively on Android 12 in this application. What do you guys think?

**2026 edit:** This discussion was converted into a UI & UX thread to discuss any improvements related to the interface and my current rewrite of the UI using Material 3 Expressive

**Comment by slowscript** [comment]
My knowledge of Material U only reaches as far as what was covered in that LTT video so I have no idea how much work and code changes it would need or how great is the risk of breaking the UI for existing users. I do like it though. Maybe I'd wait until Android 12 is released and see how other devs use it.

In the meantime, feel free to play with it if you want to. Maybe share us some screenshots :smiley:

**Comment by raresvanca** [comment]
# Material 3 Expressive re-design update in progress

For those interested in providing feedback or monitoring development progress, here is a status update on the UI rewrite:

### 1. Manual Connection & QR Optimization
The Manual Connection Dialog has been enhanced with an updated icon, new layout and significant rendering performance improvements:
*   **Optimization:** QR code generation now runs using modern Coroutines.
*   **Memory Efficiency:** We now use `Bitmap.Config.ALPHA_8`. Since QR codes only contain black/transparent pixels, this minimizes memory footprint. Color tinting is applied via a color Filter during the render pass.
*   **Visual Quality:** By reducing the render size and applying `FilterQuality.None`, the output is now crisp and pixel-perfect, avoiding the blurriness of bilinear filtering.
*   **UX:** Added "Smart Paste" logic. If the clipboard contains a valid `warpinator://` URI, it is automatically validated and added to the list. Added an empty state for when no recent remotes exist.

https://github.com/user-attachments/assets/4ad9064a-0423-4133-a92c-16b92f940dfd

<img width="353" height="768" alt="Image" src="https://github.com/user-attachments/assets/61db14fc-1df8-4f5a-bc56-763fbb783a30" /> 

<img width="349" height="399" alt="Image" src="https://github.com/user-attachments/assets/5162feb8-8a17-48fa-9770-df141bbc3e46" />

### 2. Adaptive Layouts
The home screen now implements the Material 3 [List-Detail canonical layout](https://m3.material.io/foundations/layout/canonical-layouts/list-detail). This ensures optimal functionality on tablets, foldables, and landscape modes.

https://github.com/user-attachments/assets/7dba1aaf-997f-4b68-837a-b78698b24e22

And here is the layout on mobile devices featuring the **new expandable transfer tiles**.

https://github.com/user-attachments/assets/a216aab1-5902-4616-bf9d-85aff4a25b15

<img width="347" height="752" alt="Image" src="https://github.com/user-attachments/assets/56e7c6f2-6833-44bf-9710-67347bd544d2" />

<img width="347" height="364" alt="Image" src="https://github.com/user-attachments/assets/598b1bca-ea52-4413-8168-ca09b770757c" />

### 3. Rescan & Settings page
*   **Refresh Indicator:** Added a Material 3 pull-to-refresh indicator to also trigger a rescan of remotes
*   **Settings Screen:** Now using segmented list tiles like the Settings app on Android 16
*   **Profile Picture Dialog:** Redesigned as a full-screen interface with more colors options based on a simplified version of the Material 3 tonal palette algorithm.

https://github.com/user-attachments/assets/4f9728c0-13eb-4afc-872d-9ac00574d70a

### 4. States
*   **State Feedback:** Worked on clearer visual representations for different service/app states:

<img width="841" height="610" alt="Image" src="https://github.com/user-attachments/assets/b26ba6c8-a9ea-4dbe-9ce3-b4d0254f7381" />

<img width="830" height="610" alt="Image" src="https://github.com/user-attachments/assets/d2dd4840-4a73-4a12-b743-685a4ea8cbb4" />

<img width="549" height="610" alt="Image" src="https://github.com/user-attachments/assets/6cae9d7d-e0b3-4347-8e97-61e19dc88094" />

https://github.com/user-attachments/assets/222863e9-8c06-4f7d-bf7c-9f9373691b7a

**Comment by slowscript** [comment]
Hi Rares, good to see that you are still around. As you can see, the app design stayed mostly the same as you left it, so a good chunk of the positive feedback the app receives belongs to you. I've probably only made it look worse and less consistent. If you think it could use another redesign, it is definitely welcome (although I might have my nitpicks as usual).

As you suggested, putting it into a separate branch at some point could be fine, maybe merging it part by part later on. I've looked at your repository and noticed that you've reorganized the project quite a bit. If you want, I can split the code in master branch into 'core' and 'views' subfolders to make the final diff smaller and easier to rebase on new changes. Doing more levels however feels a bit excessive to have just a couple of files per folder. Another thing is that the minSDK was increased. I personally would like to continue supporting Android 5 as long as possible, until it becomes seriously impractical to do so. I'm pretty sure Compose should support Android 5, from last time I looked into it.

Visually, most of the changes look great. The list-detail layout is something I was thinking of doing for a while but never got to it. Maybe one thing to consider is the need for a start/restart/stop screen. Ideally these states should take no time (especially stop, I think it currently is instantaneous and definitely doesn't need a screen). I think I like better the current approach where the full UI shows up with text indicating that the service is initializing in the background. For me it's around 2-5 seconds from cold start to fully operational (depending on which phone I use) and having the UI show up in the middle of those 5 seconds makes it feel more snappy than waiting full 5 seconds just looking at a rotating shape.

**Reply by raresvanca** [reply]
Thanks for the response!

> I've looked at your repository and noticed that you've reorganized the project quite a bit.

Yeah, Compose really requires splitting the UI into multiple files, so a feature-based folder structure is much cleaner than dumping everything in one place.

Ignore my repo for now, though, I have a lot of local changes I'm still organizing. I want to make sure the refactor is solid before I push, rather than force-pushing half-baked commits.

> I'm pretty sure Compose should support Android 5, from last time I looked into it.

That was the case previously, but Jetpack/AndroidX libraries enforced a `minSdk` of 23+ starting in June 2025([more info](https://developer.android.com/jetpack/androidx/versions#version-table), look a bit above the anchor). We're  blocked from Android 5 if we want to use the latest libraries (meaning Material 3 Expressive)

> Maybe one thing to consider is the need for a start/restart/stop screen

I agree that perceived performance is crucial. However, the goal of the state screens isn't to hide the UI, but to avoid **implicit states**.

If the UI appears ready but the server isn't initialized, the app is in a blocking state. Users often interpret a non-responsive "ready" UI as a bug or freeze. The new loading screen is designed to be:
* **Transient:** Ideally invisible on fast devices.
* **Informative:** Only visible if startup hangs, letting the user know the system is working rather than broken.

Trust me the screens don't show up if they don't need to. The startup is still almost instant, usually finishing before the first frame renders.

The spinner part is only how the new [indeterminate](https://m3.material.io/components/loading-indicator/overview) loading indicator looks, with one custom shape that represents the warpinator logo.

> As you suggested, putting it into a separate branch at some point could be fine

A separate branch, and a potential Alpha/Beta release would be amazing. 

Honestly, the scope grew larger than I expected. I thought I could just write a new UI on top of the old code, but the broadcasting system fought against Compose. I ended up having to move to Hilt and implement modern Kotlin features (coroutines, null safety, flows etc.) to make it work. 

Since this is now a full migration, I think it's best if I finish the heavy lifting on my side first. We can discuss merging later once I'm sure I haven't broken anything and if you want after you take a look at how the new code looks.

**Comment by raresvanca** [comment]
Demo of sharing text, editing before sharing and the share dialog as a whole:

[Phone SR.webm](https://github.com/user-attachments/assets/91bd94a0-56b6-4cc0-9539-745a4b25206a)

Demo of the messaging pane on large devices. On folding screens (half-folded or with a visible hinge), the panes divide at the hinge:

https://github.com/user-attachments/assets/789b4742-2b88-4686-bb82-cd0ca846adec

Today, I finished migrating the new text messaging system and the new adaptive share dialog (on phones is a full screen dialog, and on large devices is a floating one).

I will also add a setting to disable the messaging pane for users who prefer seeing messages as transfers in a single list rather than in a separate section.

Other screenshots:
<img width="540" height="1212" alt="Screenshot of the new About screen" src="https://github.com/user-attachments/assets/1744ffa1-bef6-4451-9b8c-b700f4a47bf9" />

![Screenshot_20260226_202311_Warpinator](https://github.com/user-attachments/assets/a32d96ae-3731-45ad-8188-1f20ae0fbe86)

**Comment by raresvanca** [comment]
I’ve officially completed the Jetpack Compose & Material 3 migration. The remaining tasks on my list are limited to string refactoring and accessibility improvements.

@slowscript I'd appreciate your feedback whenever you have a moment. Thanks!

Key updates since the last check-in:

- Added a setting to toggle Material You dynamic coloring
- Added an option to hide the extra messages panel and display messages directly within the transfers panel
- Implemented logic to suppress system notifications if a user is actively viewing a remote's transfers. It also now automatically marks those notifications as read/cleared.
- Proper ripples for the message bubbles

Here are some videos to show the final state of the UI

https://github.com/user-attachments/assets/46d606a0-dee7-44f8-8f5d-25dee604091b


https://github.com/user-attachments/assets/a9514381-cdc2-4ca7-950d-3f7d62e3f8af

**Reply by raresvanca** [reply]
Finished everything related to accessibility: from proper semantics making the app entirely usable with a screen reader to adding support for mouse, keyboards and keyboard shortcuts, drag and drop and multi window mode.

<img width="1990" height="1080" alt="Screenshot_20260311-112319-display-6" src="https://github.com/user-attachments/assets/7db29050-a804-4106-8607-ec1272d9851f" />
<img width="1990" height="1080" alt="Screenshot_20260311-112334-display-6" src="https://github.com/user-attachments/assets/083105a0-b318-4e6c-b4c6-c7fa35ee7a1b" />
<img width="1990" height="1080" alt="Screenshot_20260311-112402-display-6" src="https://github.com/user-attachments/assets/23424577-2ef9-41fc-bec8-c03ce4807d13" />

Also started experimenting with XR support, but that is irrelevant now:
<img width="2560" height="2558" alt="Screenshot_20260307_204044" src="https://github.com/user-attachments/assets/6d260526-5447-4efb-99b2-e12ef9036e1e" />
